### DISCUSSION TITLE
Improve `Fortran` syntax

### DISCUSSION THREAD

**Discussion by miEsMar** [vim/vim#13515] [General]
Hi all, 

I would like to ask for some good resources from where to learn syntax for vim's syntax files. 
I started editing `fortran.vim`, because I guess it lacks some consistency. 
Got to a decent stage, but I am missing some advanced capabilities in order to make some other things to work.

Any suggestion is more than welcome. 
Inded, after I can certainly open a PR if this solution might improve built-in `Fortran` syntax highlighting. 

Thanks

PS. This also comes because I am actually using Neovim (please, don't blame me), but the syntax that comes with Fortran treesitter is even worse than the built-in one (IMO). So, I thought that providing an excellent built-in syntax highlighting in Vim would be the best option.

**Comment by chrisbra** [comment]
please check the included documentation. In particular you might be interested in `:h pattern.txt` and `:h usr_44.txt`  and `:h syntax.txt`

**Comment by dkearns** [comment]
Have you reached out to the maintainer?  Ajit has been maintaining the Fortran runtime files for many years, decades at this point, and some of what you're hoping to fix may not be achievable.

You can learn quite a lot from reviewing other syntax files[1] to see how they approach similar problems in other languages.  There's some significant holes in the current capabilities and some de facto kludges to work around them.  Feel free to just ask specific questions.

1. The TextMate format grammar files used in VS Code can sometimes provide hints on how to approach things as well.

**Comment by miEsMar** [comment]
@chrisbra thanks for the suggestions. They halped a lot.
@dkearns thanks. Yes, I have sent him an email, to which he responded quickly. I then replied showing the proposed changes, but still didn't have answer to that.

To be clear, *my* version is highly influenced by his original (I even had an older version, coming from the Neovim installation. He kindly provided me the latest from this repo). 
I just fixed some spots here and there, and grouped by main constructs.

I'll attach screenshots comparing the current vs. proposed solution, and everyone can express his opinion. 
If it suits the majority, as said, I can make a simple PR.

### Current syntax highlighting
![immagine](https://github.com/vim/vim/assets/144672975/1580c282-abda-4351-adcd-67b2f36444d5)
![immagine](https://github.com/vim/vim/assets/144672975/d002ba44-6dae-4b5c-a03b-8c77751e3edb)

### Proposed
![immagine](https://github.com/vim/vim/assets/144672975/619d8b2c-b765-4659-adbe-8f06aac22149)
![immagine](https://github.com/vim/vim/assets/144672975/5234d216-87c9-46b5-bc66-c4a5a1431df9)


## Majour changes

1. Some language keywords are not highlighted if not in the right sematical place (see `data` in `MODULE data`)
2. Highlight procedure names (as intrinsic ones) (in all places, at their definition or within an interface block)
3. Highlight type parameters (as constant) in variable type declaration, keeping valid built-in language constructs (numbers, LK parameters, and old fashion declarations)
4. Type access modifiers, parameters, and attributes highlighted as StorageClass modifiers, not as type themselves.
5. Highlight Constant variables (which are supposed to be __ALL_UPPERCASE_LETTERS)
6. Add highlight for derived type access operator `%`
7. Fix missing keyword highlights
8. Other small fixes

Open to discussion/criticism.

Regards

**Comment by dkearns** [comment]
It seems like an improvement, thanks.  Can you please submit a PR if it's ready?  It's easier to review the changes with access to the actual syntax file.  Can you also include a plain text sample of the example source code?

It's Ajit's call to make and things can move a little slowly at times when it comes to runtime file maintenance but, irrespective, it's better to have it queued.
